proxy/http/remap/PluginDso.cc

```diff
@@ -215,6 +216,18 @@ PluginDso::modTime() const
   return _mtime;
 }
 
+/**
+ * @brief file handle returned by dlopen syscall
+ *
+ * @return dlopen filehandle
+ */
+
+void *
+PluginDso::dlOpenHandle() const
```

Comment:
Yeah, I had the same thing in mind. First, I feel it is a reasonable assumption to assume that the dso will be opened using dlopen and thus asking for its handle seems reasonable as well.

I looked at how we can use the suggestion of having a class derived from PluginDso where this member can be acessed in unit-tests, and there does not seem to be an easy way to do that without writing too much bolier-plate code for a small change. I will try to explain here.
There are calls like `RemapPluginInst *plugin_v2      = factory2->getRemapPlugin(configName, 0, nullptr, error2, isPluginDynamicReloadEnabled());` which return an object of type `RemapPluginInst` which holds a pointer to an object of type `RemapPluginInfo`. In order to get get access to `_dlh` without this `PluginDso::dlOpenHandle()` function, I think I would have to define `class RemapPluginInfoTest : public RemapPluginInfo` and then define copy constructors in class `RemapPluginInfoTest`, `RemapPluginInfo` and `PluginDso`, so that I can do something like
`RemapPluginInst *plugin_v2      = factory2->getRemapPlugin(configName, 0, nullptr, error2, isPluginDynamicReloadEnabled());`
`RemapPluginInfoTest test_object(plugin_v2->_plugin);`
`CHECK(test_object._dlh == ...)`

In short, having this function `PluginDso::dlOpenHandle()` seems fine to me. Please let me know if I missed something or if you have any further thoughts.

Reply:
not sure why my response does not show up here. Please see - https://github.com/apache/trafficserver/pull/6421#discussion_r383079718

Reply:
> First, I feel it is a reasonable assumption to assume that the dso will be opened using dlopen and thus asking for its handle seems reasonable as well.

Encapsulation can be an obstacle to simplicity but it has its benefits - think in terms of preventing the `PluginDso` user from calling `dlclose` directly, also about possibility to implement internal `PluginDso` housekeeping 

> In short, above seems like an over-kill 

I think it could be expected to have little bit more "cumbersome" unit-test code in the name of better module design. IMHO it does not mean that this is not worth pursuing, it rather means that we have not come up with a better way to do it.

> Clean and simple

True, access modifiers, design patterns often go in the opposite direction to simplicity (similarly security vs usability) but it is believed that they could help long term. Also sometimes it just seem more complex than it actually is (i.e. removing the coupling in `PluginDso` did not really make the code more complex)

> I would prefer to keep it that way.

Fair enough, I appreciate the thorough unit-tests related to the plugin isolation! I think we are in a good shape here - coverage is good and we can catch regressions in future which is the final goal here.